packages/repository/src/repositories/lyric/repository.ts

```diff
@@ -0,0 +1,135 @@
+import * as path from 'path';
+
+import {
+	type DataRecordValue,
+	type IRepository,
+	logger,
+	type LyricRepositoryConfig,
+} from '@overture-stack/maestro-common';
+
+import httpClient from '../../network/httpClient';
+import { isArrayOfObjects } from '../../utils/utils';
+
+/**
+ * Implementation to use Lyric as a data repository
+ * @param config
+ * @returns
+ */
+export const lyricRepository = (config: LyricRepositoryConfig): IRepository => {
+	const { baseUrl, categoryId, paginationSize } = config;
+	const getRepositoryRecords = async function* (): AsyncGenerator<Record<string, DataRecordValue>[], void, unknown> {
+		let page = 1;
+		let hasMoreData = true;
+
+		// Get all records paginated by category
+		// http://lyric/data/category/{categoryId}?view=compound&page={pageNumber}&pageSize={pageSize}
+		const fullUrl = new URL(path.join('data', 'category', categoryId.toString()), baseUrl);
+		fullUrl.searchParams.append('view', 'compound');
+		if (paginationSize) {
+			fullUrl.searchParams.append('pageSize', paginationSize.toString());
+		}
+		while (hasMoreData) {
+			if (paginationSize) {
+				fullUrl.searchParams.set('page', page.toString());
+			}
+			try {
+				const result = await httpClient(fullUrl.toString());
+
+				if (isArrayOfObjects(result?.records)) {
+					const validArray: Array<Record<string, DataRecordValue>> = result.records.map(
+						(item: Record<string, DataRecordValue>) => {
+							const { systemId, ...rest } = item;
+							return { ...rest, id: systemId };
+						},
+					);
+					yield validArray;
+				} else {
+					return;
+				}
+				if (paginationSize) {
+					page++;
+					hasMoreData = page < result?.pagination?.totalPages;
+				}
+				hasMoreData = false;
+			} catch (error) {
+				logger.error(`Error fetching Lyric records on category '${categoryId}'`, error);
+			}
```

Comment:
If the request always fails we will get stuck in an infinite look because `hasMoreData` is never updated.